keps/sig-api-machinery/2211-omit-managed-fields-server-side/README.md

```diff
@@ -0,0 +1,372 @@
+    if _, err := meta.Accessor(obj); err == nil {
+        obj = obj.DeepCopyObject()
+        a, _ := meta.Accessor(obj)
+        a.SetManagedFields(nil)
+    } else if meta.IsListType(obj) {
+        obj = obj.DeepCopyObject()
+        _ = meta.EachListItem(obj, func(item runtime.Object) error {
+            a, err := meta.Accessor(item)
+            if err != nil {
+                // ignore
+                return nil
+            }
+            a.SetManagedFields(nil)
+            return nil
+        })
+    }
+    return s.inner.Encode(obj, w)
+}
+
+func (s *serializer) Decode(data []byte, defaults *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
+    return s.inner.Decode(data, defaults, into)
+}
+```
+
+3. Make `k8s.io/apimachinery/pkg/runtime/serializer.newSerializersForScheme` returns serializers for the new "unmanaged" content types.
+For instance:
+```go
+import (
+    ...
+    "k8s.io/apimachinery/pkg/runtime/serializer/unmanaged"
+)
+
+func newSerializersForScheme(scheme *runtime.Scheme, mf json.MetaFactory, options CodecFactoryOptions) []serializerType {
+    ...
+    serializers := []serializerType{
+        jsonSerializerType,
+        {
+            AcceptContentTypes: []string{runtime.ContentTypeYAML},
+            ContentType:        runtime.ContentTypeYAML,
+            FileExtensions:     []string{"yaml"},
+            EncodesAsText:      true,
+            Serializer:         yamlSerializer,
+        },
+        {
+            AcceptContentTypes: []string{runtime.ContentTypeProtobuf},
+            ContentType:        runtime.ContentTypeProtobuf,
+            FileExtensions:     []string{"pb"},
+            Serializer:         protoSerializer,
+
+            Framer:           protobuf.LengthDelimitedFramer,
+            StreamSerializer: protoRawSerializer,
+        },
+    }
+    ...
+
+    allSerializers := append([]serializerType{}, serializers...)
+    for _, serializer := range serializers {
+        serializer.ContentType += runtime.ContentTypeSuffixOmitManagedFields
+        for i := range serializer.AcceptContentTypes {
+            serializer.AcceptContentTypes[i] += runtime.ContentTypeSuffixOmitManagedFields
+        }
+        serializer.Serializer = unmanaged.NewSerializer(serializer.Serializer)
+        serializer.StreamSerializer = unmanaged.NewSerializer(serializer.StreamSerializer)
+        serializer.PrettySerializer = unmanaged.NewSerializer(serializer.PrettySerializer)
+        allSerializers = append(allSerializers, serializer)
+    }
+    return allSerializers
+}
+```
+
+### Client-side
+
+<<[UNRESOLVED @knight42 ]>>
+I am uncertain about what the flag should look like in kubectl.
+<<[/UNRESOLVED]>>
+
+#### Alternative 1: New dedicated flag, for example `--show-managed-fields`
+
+The value of flag `--show-managed-fields` could be `omit`, `default` and a version name(in case users want to 
```

Comment:
I strongly prefer this alternative because it could be defaulted in a kubeconfig file.

I'd name the flag `--managed-fields-version={omit|v1}` or possibly just `--managed-fields={omit|v1}`.

Reply:
cc @pwittrock any opinion?

Reply:
Where is this flag? kubectl only?

Reply:
:+1: for sig-cli pov, it allows much more flexibility for the future.

Reply:
I'm even inclined to have it default to `omit`.

Reply:
I agree with @lavalamp 
Having the flag named `--show-managed-fields` indicates a boolean value and the options presented here don't match that.